### DISCUSSION TITLE
New slot model "SlotW31B"

### DISCUSSION THREAD

**Discussion by zmladen** [Eomys/pyleecan#711] [General]
Hello,

I decided to try Pyleecan for a project I'm working on. My first impression is that it's a well-structured and thoughtfully designed project with a very flexible framework. However, due to the wide range of options, it took me a little time to get started. To some of its capabilities I decided to create a new slot model. This is the schematics of the resulting slot i've created with pyleecan:

![image](https://github.com/user-attachments/assets/17e09e26-9684-4c90-9764-d2c73aa4f228)

The model I created differs from the existing ones by incorporating isolation gaps between the active part of the winding and the lamination slot. Additionally, it allows for the inclusion of gaps between tangential divisions in the winding layout. 

The next thing i would like to implement is the visualization of the wires inside the active areas like here:

![image](https://github.com/user-attachments/assets/c8ccc3d3-08dd-44a6-8395-a670d4a3756a)

Any thoughs on how to implement it in the best and flexible way are welcome. 

Now I would like to contribute this new slot to the Pyleecan project. Could someone guide me through the process of contributing, or provide support in adding / testing this model to the pyleecan project?

I have creates a fork of the project:

[https://github.com/zmladen/pyleecan](url)

Thanks,
zmladen

**Comment by BonneelP** [comment]
Hello,

Thank you for your interest in pyleecan !

Adding a new slot is a good way to start with pyleecan as it can help you to discover several key aspects of the development (class generator, PR, Tests...). We have a tutorial about how to add a slot in pyleecan that can help you to contribute: https://www.pyleecan.org/tuto.add.slot.html but don't hesitate to ask questions if you need help in creating the PR, the tests, the GUI... By the way, you may need to do something different than usual for get_surface_opening to avoid that the airgap mesh size goes inside your slot. 

Regarding the visualization of the wires this would be a great addition to pyleecan ! For that I would say the best way would be to add a new generic method to Surface abstract object that would return a list of Circle object corresponding to the conductors. But then we would need to find a algorithm to properly add the circle in any surface which is tricky ! Do you have an idea of how we could do that or do you plan to do a method that works only on your schematics (which would be a great step any way) ? Something interesting is that the geometry modeler is central in pyleecan, so if you manage to find a way to plot it, we will be able to simulate that without much effort in a FEA software. 

Best regards,
Pierre

**Reply by zmladen** [reply]
Hello Pierre,

Thanks for the quick response. So far, I'm really enjoying the project and I plan to contribute more in the future. About your comment regarding get_surface_opening, is it primarily for controlling the mesh? Is the entire area marked in yellow needed to generate surface opening part?

![Unbenannt1](https://github.com/user-attachments/assets/198130d5-884e-4f40-a495-11fe341bc9f6)

Can I add a cut in the middle like in the picture bellow and add all the points that surround the surface opening?

![image](https://github.com/user-attachments/assets/e872e5ca-c5a3-41af-a5d2-c5554be26c6c)

Please let me know what it the best way to do it as i still need some learning time for pyleecan.

**Reply by zmladen** [reply]
Can I break down the isolation surface into smaller regions and combine them in single slot opening surface? Which methods are the best to use for that? If someone can share any example for that it would be great.

**Reply by BonneelP** [reply]
Hello,

Indeed, "opening surface" is only for mesh purpose and yes it's all the yellow one (all the "air" in the slot). You can return several surfaces so I will be easier to split it in several smaller surface (to define a SurfLine instead of a SurfRing which is the surface between an outer and an inner surface like a lamination). You will need to adapt these surface if i_tan>0. 

Best regards,
Pierre

**Comment by zmladen** [comment]
Hello Pierre,

i imagined it would be so. I have added 8 surfaces as shown on the picture:

![image](https://github.com/user-attachments/assets/5090a067-d282-4aaf-8b9a-9dcd65854f7e)

This only works if all gaps are greater than zero (i_bore > 0, i_tooth > 0, i_yoke > 0 and i_tan>0). I believe this will always hold true for this slot. If needed, for the slots without gaps I would make new slot class. the mentioned constrains (i_bore > 0, i_tooth > 0, i_yoke > 0 and i_tan>0) I will add in the slot class’s check function.

Are there any constraints regarding the direction of the lines between neighboring surfaces? 

![image](https://github.com/user-attachments/assets/1ca1a873-310a-4c48-9a00-9b94b2c3591b)

**Reply by BonneelP** [reply]
Hello,

I think you can add some conditions in get_surface_opening to change the number of returned surfaces if some parameters are at 0. We had a lot of surprises with how some users have used the schematics, so always assume the worst ^^" You can check what appends if any parameters is at 0, several at 0 at the same time or if some implicit constraint are broken (i_yoke, i_tooth... larger than the slot)

There is no particular constraint on the lines directions.

Best regards,
Pierre

**Reply by zmladen** [reply]
Thanks Pierre, i will do this for the case i_tan = 0. Other cases I will add later if needed. Before that I would like to test if the slot can be used in the simulation.

Regards,
Mladen
